go/vt/vtctl/reparentutil/planned_reparenter.go

```diff
@@ -0,0 +1,627 @@
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+)
+
+// PlannedReparenter performs PlannedReparentShard operations.
+type PlannedReparenter struct {
+	ts     *topo.Server
+	tmc    tmclient.TabletManagerClient
+	logger logutil.Logger
+}
+
+// PlannedReparentOptions provides optional parameters to PlannedReparentShard
+// operations. Options are passed by value, so it is safe for callers to mutate
+// resue options structs for multiple calls.
+type PlannedReparentOptions struct {
+	NewPrimaryAlias     *topodatapb.TabletAlias
+	AvoidPrimaryAlias   *topodatapb.TabletAlias
+	WaitReplicasTimeout time.Duration
+
+	// Private options managed internally. We use value-passing semantics to
+	// set these options inside a PlannedReparent without leaking these details
+	// back out to the caller.
+
+	lockAction string
+}
+
+// NewPlannedReparenter returns a new PlannedReparenter object, ready to perform
+// PlannedReparentShard operations using the given topo.Server,
+// TabletManagerClient, and logger.
+//
+// Providing a nil logger instance is allowed.
+func NewPlannedReparenter(ts *topo.Server, tmc tmclient.TabletManagerClient, logger logutil.Logger) *PlannedReparenter {
+	pr := PlannedReparenter{
+		ts:     ts,
+		tmc:    tmc,
+		logger: logger,
+	}
+
+	if pr.logger == nil {
+		// Create a no-op logger so we can call functions on pr.logger without
+		// needing to constantly check it for non-nil first.
+		pr.logger = logutil.NewCallbackLogger(func(e *logutilpb.Event) {})
+	}
+
+	return &pr
+}
+
+// ReparentShard performs a PlannedReparentShard operation on the given keyspace
+// and shard. It will make the provided tablet the primary for the shard, when
+// both the current and desired primary are reachable and in a good state.
+func (pr *PlannedReparenter) ReparentShard(ctx context.Context, keyspace string, shard string, opts PlannedReparentOptions) (*events.Reparent, error) {
+	opts.lockAction = pr.getLockAction(opts)
+
+	ctx, unlock, err := pr.ts.LockShard(ctx, keyspace, shard, opts.lockAction)
+	if err != nil {
+		return nil, err
+	}
+
+	defer unlock(&err)
+
+	if opts.NewPrimaryAlias == nil && opts.AvoidPrimaryAlias == nil {
+		shardInfo, err := pr.ts.GetShard(ctx, keyspace, shard)
+		if err != nil {
+			return nil, err
+		}
+
+		opts.AvoidPrimaryAlias = shardInfo.MasterAlias
+	}
+
+	ev := &events.Reparent{}
+	defer func() {
+		switch err {
+		case nil:
+			event.DispatchUpdate(ev, "finished PlannedReparentShard")
+		default:
+			event.DispatchUpdate(ev, "failed PlannedReparentShard: "+err.Error())
+		}
+	}()
+
+	err = pr.reparentShardLocked(ctx, ev, keyspace, shard, opts)
+
+	return ev, err
+}
+
+func (pr *PlannedReparenter) getLockAction(opts PlannedReparentOptions) string {
+	return fmt.Sprintf(
+		"PlannedReparentShard(%v, AvoidPrimary = %v)",
+		topoproto.TabletAliasString(opts.NewPrimaryAlias),
+		topoproto.TabletAliasString(opts.AvoidPrimaryAlias),
+	)
+}
+
+// prelightChecks checks some invariants that pr.reparentShardLocked() depends
```

Comment:
typo: preflightChecks

Reply:
"prelight" 😂 

nice catch, should be fixed, and i've cleaned up the commit history as well now